ush/python/pygw/src/pygw/argsparse.py

```diff
@@ -0,0 +1,210 @@
+    args_obj: object
+
+        A Python object containing the command-line arguments; if no
+        command-line arguments have been passed to the caller script,
+        the returned value is NoneType.
+
+    """
+
+    def __check__(self, args_obj: AttrDict) -> AttrDict:
+        """
+        Description
+        -----------
+
+        This method checks that the argument values are formatted
+        correctly; this method is necessary due to the Python
+        interpretter assuming that all arguments passed via the Python
+        `sys` module are formatted as strings.
+
+        Parameters
+        ----------
+
+        in_dict: object
+
+            A Python object containing the argument values to be
+            evaluated and formatted as neccessary.
+
+        Returns
+        -------
+
+        out_dict: object
+
+            A Python object containing the formatted argument values.
+
+        """
+
+        # Define local function to sort and format the input Python
+        # dictionary upon entry.
+        def sorted_by_keys(in_dict):
+
+            out_dict = AttrDict()
+            for (key, value) in sorted(in_dict.items(), key=lambda key: key):
+
+                if isinstance(value, dict):
+                    setattr(out_dict, key, sorted_by_keys(value))
+
+                else:
+                    test_value = value
+
+                    # Check if the key and value pair is a boolean
+                    # type argument and proceed accordingly.
+                    if isinstance(test_value, bool):
+                        if test_value:
+                            value = True
+
+                        if not test_value:
+                            value = False
+
+                    # Check if the key and value pair is a string type
+                    # argument and proceed accordingly.
+                    if isinstance(test_value, str):
+                        try:
+                            dummy = float(test_value)
+                            if "." in test_value:
+                                value = float(test_value)
+                            else:
+                                value = int(test_value)
+
+                        except ValueError:
+                            if test_value.lower() == "none":
+                                value = None
+
+                            elif test_value.lower() == "true":
+                                value = True
+
+                            elif test_value.lower() == "false":
+                                value = False
+
+                            else:
+                                value = str(test_value)
+
+                    # Update the output dictionary key and value pair.
+                    setattr(out_dict, key, value)
+
+            return out_dict
+
+        # Define the formatted output dictionary.
+        out_dict = sorted_by_keys(in_dict=args_obj)
+
+        return out_dict
+
+    def run(self) -> Union[None, AttrDict]:
```

Comment:
Just curious, what does this provide that `argparse` does not? I albeit am not an `argparse` expert but it seems like it is doing basically the same thing where:
`python thing.py --arg1=foo`
```
print(args.arg1)
foo
```
right?

Reply:
@CoryMartin-NOAA Yes, you are almost right. This functionality provides a wrapper over argparse. For example, if you had a script and wanted to add arguments to it using the traditional methodology, you would need to have a `parser.get_argument` for each variable that you wanted to pass.

This functionality negates that. As long as the `ArgsParse` object is invoked, it will collect all command line arguments. The user would then no longer need to explicitly define a `parser.get_argument` for each argument they wish to pass to a given script. They could pass anything that they want and the `ArgsParse` will format and and pass it to the respective task.

The advantage of this methodology is that it keeps the driver/calling scripts clean and does not the require a user/developer to need to add cumbersome directives to the script. 

I apologize that this wasn't more clear in my original description.

Reply:
```python
from pygw.argsparse import ArgsParse

ap = ArgsParse().run()
print(ap)
```

```bash
$> python3 thing.py  --yaml 'myyaml.yaml'
2023-03-27 17:23:47,168 - INFO     - root        : Command line argument yaml is myyaml.yaml
{'yaml': 'myyaml.yaml'}
```

However,

```bash
$> python3 thing.py  'mode' --yaml 'myyaml.yaml'
2023-03-27 17:23:37,121 - INFO     - root        : Command line argument mode is --yaml
Traceback (most recent call last):
  File "/Users/rmahajan/scratch/gwWork/reviews/global-workflow/thing.py", line 7, in <module>
    ap = ArgsParse().run()
         ^^^^^^^^^^^^^^^^^
  File "/Users/rmahajan/scratch/gwWork/reviews/global-workflow/ush/python/pygw/src/pygw/argsparse.py", line 201, in run
    msg = f"Command line argument {arg_keys[idx]} is {arg_values[idx]}"
                                                      ~~~~~~~~~~^^^^^
IndexError: list index out of range
```

`ArgsParse` can only parse named command line arguments.


I am with @CoryMartin-NOAA.  This class provides some "convenience" over the standard invocation of `argparse.add_argument` and parser.parse_args()`,  but it does so without providing a better command-line interface.  For e.g.

```bash
$> python3 thing.py  --yaml 'myyaml.yaml' -h
usage: thing.py [-h]

options:
  -h, --help  show this help message and exit
```

Whereas, if the script `thing.py` needed or had options for `--yaml`, the developer would have had to define them explicitly.

I would argue that in the attempt to make it easy to pass named command line arguments, the API has been dangerously degraded.